### PR TITLE
DOC: Add cache busting to all static assets

### PR DESCRIPTION
**Pull request by QuLogic**
## PR Summary

We have seen both in `/stable` and `/3.6.0` (e.g., #23760), some styling is broken because old CSS is cached. CSS might change from updating sphinx-gallery, mpl-sphinx-theme, pydata-sphinx-theme, etc. Adding a versioned query breaks the cache.

It's a bit over-eager to base it on Matplotlib version and not the file contents (since those dependencies may not have updated), but this should work well enough.

## PR Checklist

**Tests and Styling**
- [n/a] Has pytest style unit tests (and `pytest` passes).
- [x] Is [Flake 8](https://flake8.pycqa.org/en/latest/) compliant (install `flake8-docstrings` and run `flake8 --docstring-convention=all`).

**Documentation**
- [n/a] New features are documented, with examples if plot related.
- [n/a] New features have an entry in `doc/users/next_whats_new/` (follow instructions in README.rst there).
- [n/a] API changes documented in `doc/api/next_api_changes/` (follow instructions in README.rst there).
- [x] Documentation is sphinx and numpydoc compliant (the docs should [build](https://matplotlib.org/devel/documenting_mpl.html#building-the-docs) without error).

### CODE REVIEW

**Review by tacaswell — APPROVED** [review]

File: doc/conf.py [hunk 1]
```diff
@@ -649,3 +672,15 @@ def linkcode_resolve(domain, info):
                 f"/{tag}/lib/{fn}{linespec}")
 else:
     extensions.append('sphinx.ext.viewcode')
+
+
+# -----------------------------------------------------------------------------
+# Sphinx setup
+# -----------------------------------------------------------------------------
+def setup(app):
+    if any(st in version for st in ('post', 'alpha', 'beta')):
```
Comment (greglucas) [inline doc/conf.py:681]: I think "post" would still be considered release?

**Review by greglucas — APPROVED** [review]
minor question about "dev" vs "rel" releases. Feel free to merge if that is irrelevant here.

File: doc/conf.py (hunk 1 above)
Comment (QuLogic) [reply doc/conf.py:681]: That's a good question, but I just moved this code to the bottom of the file. It looks like that's how it was when it came in, https://github.com/matplotlib/matplotlib/pull/13870/commits/310b004682551644bdcb90de9e3f9bc13974a1eb so maybe @tacaswell knows why

Comment (tacaswell) [reply doc/conf.py:681]: I think we used to use `.postN` to count the number of commits past the last tag, but we switch to using. `.dev` for the number of commits on a pre-release of our expected next tag...but I think that means we need to add `'dev'` to this list as well?

Comment (tacaswell) [inline doc/conf.py:681]:
```suggestion
    if any(st in version for st in ('post', 'dev', 'alpha', 'beta')):
```

**Comment by tacaswell** [comment]
@QuLogic can self-merge when CI is green.

**Pull request merged**